include/uk/arch/spinlock.h

```diff
@@ -30,22 +29,30 @@
 extern "C" {
 #endif
 
-typedef struct {} spinlock_t;
+#include <uk/arch/lcpu.h>
 
-#ifdef CONFIG_SMP
-#error "Define your spinlock operations!"
-#else
-
-#define ukarch_spin_lock_init(lock)      (void)(lock)
-#define ukarch_spin_is_locked(lock)      (void)(lock)
-#define ukarch_spin_lock(lock)           (void)(lock)
-#define ukarch_spin_trylock(lock)        (void)(lock)
-#define ukarch_spin_unlock(lock)         (void)(lock)
-
-/* Defines a preinitialized spin_lock in unlocked state */
-#define DEFINE_SPINLOCK(lock)            spinlock_t lock = {}
-
-#endif
+#ifdef CONFIG_HAVE_SMP
+#include <uk/asm/spinlock.h>
+#else /* CONFIG_HAVE_SMP */
+/* N.B.: For singlecore systems we remove spinlocks by mapping functions to
+ * void operations. No state is saved for the spinlock but we assume that the
+ * CPU is always able to get the lock. Trying to acquire a spinlock thus always
+ * succeeds and asserting that a lock is held returns false.
+ */
+typedef struct __spinlock {
+	/* empty */
+} __spinlock;
+
+#define UKARCH_SPINLOCK_INITIALIZER()
```

Comment:
This won't compile when CONFIG_HAVE_SMP is not enabled:
```
spinlock_t mylock = UK_SPINLOCK_INITIALIZER();
```

Reply:
```suggestion
#define UKARCH_SPINLOCK_INITIALIZER()	{}
```

yep, good catch!